src/improc.c

```diff
@@ -2512,3 +2512,175 @@ DoubleImage ifft2DDouble(ComplexImage image) {
 
   return im;
 }
+
+typedef struct quack {
+  int *buffer;
+  int start;
+  int end;
+  int size;
+  int capacity;
+} Quack;
+
+inline int quackIsEmpty(Quack *quack) {
+  return quack->size == 0;
+}
+
+inline int quackPeekBack(Quack *quack) {
+  return quack->buffer[(quack->end + quack->capacity - 1) % quack->capacity];
+}
+
+static int quackPopBack(Quack *quack) {
+  int value = quackPeekBack(quack);
+  quack->end = (quack->end + quack->capacity - 1) % quack->capacity;
+  quack->size--;
+  return value;
+}
+
+static void quackPushBack(Quack *quack, int value) {
+  if (quack->size >= quack->capacity) {
+    fatalError("attempted to insert into full quack. capactiy: %d", quack->capacity);
+  }
+  quack->buffer[quack->end] = value;
+  quack->end = (quack->end + 1) % quack->capacity;
+  quack->size++;
+}
+
+inline int quackPeekFront(Quack *quack) {
+  return quack->buffer[quack->start];
+}
+
+static int quackPopFront(Quack *quack) {
+  int value = quackPeekFront(quack);
+  quack->start = (quack->start + 1) % quack->capacity;
+  quack->size--;
+  return value;
+}
+
+#if 0
+// this function is unused, but i don't know if we want to keep it for completion sake
+static void quackPushFront(Quack *quack, int value) {
+  if (quack->size >= quack->capacity) {
+    fatalError("attempted to insert into full quack. capactiy: %d", quack->capacity);
+  }
+  quack->start = (quack->start + quack->capacity - 1) % quack->capacity;
+  quack->buffer[quack->start] = value;
+  quack->size++;
+}
+#endif
+
+static Quack createNewQuackWithMemory(int capacity, int *memory) {
+  Quack quack;
+  quack.buffer = memory;
+  quack.start = 0;
+  quack.end = 0;
+  quack.size = 0;
+  quack.capacity = capacity;
+
+  return quack;
+}
+
+/**
+* Compute a sliding window (of size `w`) minimum/maximum depending on the `ord` flag of the 1D array (size = n).
+* The `offset` and `start` parameters are used to change the indexing in the array so that element i is accessed at
+* location i * offset + start. A `workspace` array (of size w elements) is given (does not need to be initialised) to
+* use as memory for the quack implementation.
+*
+* @param img The 1D int array that the sliding window operation will be computed on
+* @param out The 1D int array that the result of the operation will be stored in
+* @param n The size of both of the arrays (in nr of elements)
+* @param w The size of the window
+* @param ord If `ord` == 1, the sliding minimum is computed. Otherwise a sliding maximum is computed
+* @param offset Determines the gap in position in the input/output arrays between two successive elements
+* @param start Determines the position of the first element in the input/output arrays
+* @param workspace An array of size w (in nr of elements) to use as backing storage for our internal datastructure.
+*/
+static void slidingWindowOrd(int *img, int *out, int n, int w, int ord, int offset, int start, int *workspace) {
```

Comment:
Someone very smart (wim hesselink I think) proved that this algorithm runs in O(n) and supposedly each element gets accessed very few times (amortized O(1) times), but it could still have considerable overhead with the ridiculous accessing that I do..

We could change the quack to both store a value and an index which would be more efficient access-wise, but more annoying to access. It could be worth it to create two local variables in this function `first` and `last`, which will be assigned to `img[quackPeekFront]` and `quackPeekBack` values (after the popping and pushing is done). These variables change infrequently, and will then be stored in a register instead. These variables are the only things from the quack that are actually compared against. Could be more efficient to do that maybe? I can't really be bothered though xD

Reply:
We can keep it like this for now and possibly change it later. We'll have to do some benchmarking in that case.

Reply:
Let's merge this for now and improve it later